rfcs/framework-0035-error-handling-and-response.md

```diff
@@ -0,0 +1,323 @@
+
+**View error handling**
+
+A Flask view can simply catch exceptions, and render a response. This can either be done per view or via special decorators.
+
+```python
+@app.route('/')
+def view1():
+    try:
+        # ...
+    except Exception:
+        return 400, ...
+```
+
+The view layer error handling is best suited for very custom error handling from the service layer. E.g implementation of a custom API with custom reposnses.
+
+The issue with view layer error handling is that only errors from the view and service/data source layer is being caught. Any error in the Flask layer (routing, request processing functions etc) will escape the try/except-clause (because they happen outside the function).
+
+Example of a 404 error:
+
+1. A routing error will result in a 404 error from the Flask layer.
+2. An object that doesn't exists in the database, will likely result in a 404 error from the service layer.
+
+A view layer can only catch the latter. Thus, if we render an error response in the view, we still have to deal with error response for Flask layer errors.
+
+**Flask's error handling/rendering**
+
+Flask and Werkzeug's request dispatching system will catch all errors happening in the three layers described. The error handling is pretty sophisticated, and is explained below.
+
+First, Flask will look for a user-provided error handler. If found, the handler is responsible for making a response. If not found, but the exception is an HTTPException (or subclass thereof), the HTTPException will simply be returned, because an HTTPException can self-render. If it's not an HTTPException, the error is simply reraised and results in an Internal Server Error. The following pseudo-code is an example of the handling:
+
+```python
+def handle_exception(e):
+    # If it's an HTTPException, use different method:
+    if isinstance(e, HTTPException):
+        return handle_http_exception(e)
+
+    handler = find_error_handler(e)
+    if handler is None:
+        # No handler was found, and it's not and HTTPException.
+        raise # <----- reraise
+    # A handler was found, use it to return response
+    return handler(e)
+
+def handle_http_exception(e):
+    handler = find_error_handler(e)
+    if handler is None:
+        # No handler was found, but it is an HTTPException.
+        return e # <----- return
+    # A handler was found, use it to return response
+    return handler(e)
+```
+
+The key in understanding the difference between an normal exception and an HTTPException is that an HTTPException is considered a valid response by Flask's ``make_reponse``. This allows it to be returned by the ``handle_http_exception()`` function.
+
+**Recap**
+
+Source of errors (exceptions raised) is three layers (Flask, View, Service). For handling these errors (i.e. making an HTTP error response) we can use Flask and Werkzeug's error handling system. The view layer is not well-suited for this.
+
```

Comment:
∑(O_O;) I think I am not quite convinced that the view layer is not well-suited... Actually, what is puzzling me is that we seem to try to build up the argument for why to use Flask's `error_handler` functionality, but handling errors via a `try-catch` in the decorators of flask-resources or via `error_handler` of Flask (functionality in flask-resources as well) is actually equivalent (or at least the difference hasn't been made clear to me... :sweat_smile: ). In both cases we are placed at the blueprint level...

Reply:
It's not equivalent because Flask-level errors escape the view layer.

Reply:
What this means is basically you'll need both Flask errors handlers and you view error. Instead, you can simply let the Flask error handler do the entire job and not duplicate error handling.